File: 1715-split-a-string-into-the-max-number-of-unique-substrings/split-a-string-into-the-max-number-of-unique-substrings.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        def f(i:int, n:int,s:str,st:set)->int:
            if i==n:
                return 0
            ans=0
            for j in range(i+1,n+1):
                temp=s[i:j]
                if temp not in st:
                    st.add(temp)
                    ans=max(ans,1+f(j,n,s,st))
                    st.remove(temp)
            return ans
        st=set()
        return f(0,len(s),s,st)

File: 1715-split-a-string-into-the-max-number-of-unique-substrings/split-a-string-into-the-max-number-of-unique-substrings.cpp
class Solution {
public:
    int f(int i,int n,string s,unordered_set<string>& st)
    {
        if(i==n)
        return 0;
        string temp="";
        int ans=0;
        for(int j=i;j<n;j++)
        {
            temp+=s[j];
            if(st.find(temp)==st.end())
            {
                st.insert(temp);
                ans=max(ans,1+f(j+1,n,s,st));
                st.erase(temp);
            }
        }
        return ans;
    }
    int maxUniqueSplit(string s) {
        unordered_set<string>st;
        int n=s.size();
        return f(0,n,s,st);
    }
};